### PR TITLE
Save webp

### PR DESCRIPTION
**Pull request by docterd**
I added the option "PINRemoteImageManagerSaveProcessedImageAsWEBP" to save processed images as WebP (currently lossy in quality 75) instead of PNG. If needed we can change it to lossy or provide a option to choose (like the jpg option).

### CODE REVIEW

File: Pod/Classes/Categories/PINImage+WebP.m [hunk 1]
```diff
@@ -77,6 +78,62 @@ + (PINImage *)pin_imageWithWebPData:(NSData *)webPData
...
+    // convert color space.
+    CGRect imageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
+    
+    NSUInteger bytesPerPixel = 4;
+    NSUInteger bytesPerRow = bytesPerPixel * image.size.width;
+    NSUInteger bitsPerComponent = 8;
+    NSUInteger bitmapByteCount = bytesPerRow * image.size.height;
+    
+    unsigned char *rawData = (unsigned char*) calloc(bitmapByteCount, sizeof(unsigned char));
```
Comment (garrettmoon) [inline Pod/Classes/Categories/PINImage+WebP.m]: Is this ever freed?

File: Pod/Classes/Categories/PINImage+WebP.m [hunk 2]
```diff
@@ -77,6 +78,62 @@ + (PINImage *)pin_imageWithWebPData:(NSData *)webPData
...
+    
+    CGDataProviderRef dataProvider = CGImageGetDataProvider(imageRef);
+    CFDataRef dataRef = CGDataProviderCopyData(dataProvider);
+    
+    picture.colorspace = WEBP_YUV420A;
+    picture.width = image.size.width;
+    picture.height = image.size.height;
+    
+    WebPPictureImportRGBA(&picture, (uint8_t *)CFDataGetBytePtr(dataRef), (int) CGImageGetBytesPerRow(imageRef));
+    WebPPictureARGBToYUVA(&picture, picture.colorspace);
```
Comment (garrettmoon) [inline Pod/Classes/Categories/PINImage+WebP.m]: I don't know much about webp, but is there an advantage to converting to YUVA?

File: Pod/Classes/Categories/PINImage+WebP.m [hunk 3]
```diff
@@ -77,6 +78,62 @@ + (PINImage *)pin_imageWithWebPData:(NSData *)webPData
     return nil;
 }
 
++ (NSData *)pin_DataFromWebPimage:(PINImage *)image {
+
```
Comment (garrettmoon) [inline Pod/Classes/Categories/PINImage+WebP.m]: Style nit: mind removing the newlines at the beginning and end of this method and putting the brace on a new line?

**Comment by garrettmoon** [comment]
CI (sorry it's internal) failed with:
    - ERROR | [PINRemoteImage/Core, PINRemoteImage/iOS, PINRemoteImage/OSX, and more...] xcodebuild:  /usr/local/var/buildkite-agent/builds/BuildKites-Mac.local-1/pinterest/pinremoteimage/Pod/Classes/PINRemoteImageManager.m:719:46: error: no known class method for selector 'pin_DataFromWebPimage:'

File: Pod/Classes/Categories/PINImage+WebP.m (hunk 3 above)
Comment (docterd) [inline Pod/Classes/Categories/PINImage+WebP.m]: No problem :-)

File: Pod/Classes/Categories/PINImage+WebP.m (hunk 1 above)
Comment (docterd) [inline Pod/Classes/Categories/PINImage+WebP.m]: Actually not. Will add it.

File: Pod/Classes/Categories/PINImage+WebP.m (hunk 2 above)
Comment (docterd) [inline Pod/Classes/Categories/PINImage+WebP.m]: I think you are right in this case. Especially I had to convert the color space anyway to workaround some color issues. I'll check if I can remove it.

**Comment by docterd** [comment]
> Could you write a test to make sure this continues to work correctly?

I'll add some when I got the time.

> How do you imagine clients to change the webp quality if they don't want it to be 75?

I think it would make more sense to keep the api simple and move from to a lossless compression (similar like the other image options).

**Comment by garrettmoon** [comment]
@docterd here's the failure:
```
Testing CocoaPod configurations
 
 -> PINRemoteImage (3.0.0-beta.7)
    - ERROR | [PINRemoteImage/Core, PINRemoteImage/iOS, PINRemoteImage/OSX, and more...] xcodebuild: Returned an unsuccessful exit code. You can use `--verbose` for more information.
    - ERROR | [PINRemoteImage/Core, PINRemoteImage/iOS, PINRemoteImage/OSX, and more...] xcodebuild:  /usr/local/var/buildkite-agent/builds/BuildKites-Mac.local-1/pinterest/pinremoteimage/Pod/Classes/PINRemoteImageManager.m:715:99: error: use of undeclared identifier 'PIN_WEBP'
```

BTW, you can run the tests that the CI runs by running `./build.sh all` in the root directory.

**Comment by garrettmoon** [comment]
Hey, @docterd would you mind rebasing your PR? The build script has been updated since you put this up.

**Comment by aureliojuarezcortez** [comment]
[GITHUB.COM](@AurelioJuarez2avatar)

**Comment by docterd** [comment]
Unfortunately I won't be able to do it in the next days, because I just went on vacation. I will adjust everything when I'm back :-/

**Comment by garrettmoon** [comment]
I'm going to close this for now, if you get the time to rebase, please put up a new PR!

**Pull request closed without merging**